### PR TITLE
GPG-669 Added autoscaling policy

### PR DESCRIPTION
**Pull request by malinaslevoaca1**
Since CPU Utilization isn't reliable I think we should use `throughput` as the autoscaling metric.
I had a look at the metrics in Grafana and 100 should be a good threshold (I _think_ we can also increase it)

**Note:** A script needs to be added to the deployment pipeline as well. I plan on using the same policy.

I haven't tested it yet.

### CODE REVIEW

**Comment by Binney** [comment]
Nice! Is this just for the ipdeny component or will we be using the same policy json for the C# server too?

File: Infrastructure/gpg-ipdeny/autoscaling_policy.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿{
+  "instance_min_count": 1,
```
Comment (Binney) [inline Infrastructure/gpg-ipdeny/autoscaling_policy.json:2]: I would expect us to have different min_count and max_counts on each environment; is that feasible? how would we go about this?

**Review by Binney — APPROVED** [review]
Nice, looking forward to seeing the test :D

**Comment by malinaslevoaca1** [comment]
> Nice! Is this just for the ipdeny component or will we be using the same policy json for the C# server too?

I'll use the same policy and we can update them later if needed

**Review by dulciejackson — APPROVED** [review]
Looks good! 
Like Sarah said we might want to have different criteria/max instance limits for the two apps, but we can look into that more once we've tested how this reacts during a load test 🙂

File: Infrastructure/gpg-ipdeny/autoscaling_policy.json (hunk 1 above)
Comment (malinaslevoaca1) [reply Infrastructure/gpg-ipdeny/autoscaling_policy.json]: Good point! I updated the script to build the json policy

File: Infrastructure/gpg-ipdeny/deploy.sh [hunk 2]
```diff
@@ -77,3 +91,10 @@ fi
 PROTECTED_APP_HOSTNAME="$(cf curl "v3/apps/$(cf app "${PROTECTED_APP_NAME}" --guid)/routes" | jq -r --arg APPS_DOMAIN "${APPS_DOMAIN}" '[.resources[] | select(.url | endswith($APPS_DOMAIN))][0].host')"
 
 cf bind-route-service "${APPS_DOMAIN}" "${ROUTE_SERVICE_NAME}" --hostname "${PROTECTED_APP_HOSTNAME}";
+
+# Autoscaling
+echo "{\"instance_min_count\":${MIN_COUNT_INSTANCES},\"instance_max_count\":${MAX_COUNT_INSTANCES},\"scaling_rules\":[{\"metric_type\":\"throughput\",\"breach_duration_secs\":60,\"threshold\":100,\"operator\":\"<\",\"cool_down_secs\":60,\"adjustment\":\"-1\"},{\"metric_type\":\"throughput\",\"breach_duration_secs\":60,\"threshold\":100,\"operator\":\">=\",\"cool_down_secs\":60,\"adjustment\":\"+1\"}]}" > autoscaling_policy.json
```
Comment (Binney) [inline Infrastructure/gpg-ipdeny/deploy.sh:96]: Is there a way of doing this which doesn't result in overwriting the file that's already in source control? e.g. putting some placeholders in the JSON and then here in the bash script you just read the file into a variable and replace the placeholders

**Review by Binney — APPROVED** [review]

File: Infrastructure/gpg-ipdeny/deploy.sh (hunk 2 above)
Comment (malinaslevoaca1) [reply Infrastructure/gpg-ipdeny/deploy.sh:96]: I removed the file but I agree - that's a better idea 👍

**Pull request merged**